# Bismark Visualisation Toolkit

[Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/) is a tool
used for aligning Bisfulfite-Sequencing libraries, giving information about
DNA methylation.

This script reads in coverage reports generated by Bismark and calculates
a number of metrics:

* Spearman's rank correlation coefficients between each sample
* Principal component analysis (PCA) eigenvectors / eigenvalues

It also generates a number of plots:

* Dendrogram of hierarchical clustering between samples
* Scatter plot for each sample pair
* Histogram of methylation percentages for each sample
* Heatmap of sample correlation scores
* Scatter plot of first two principal components
* Coverage analysis
  * Includes analysis of region of interest, if annotation supplied

## Example Output

#### Clustering Dendrogram
![Clustering Dendrogram](../../examples/bismark_dendrogram.png)

#### Correlation Score Heatmap
_Example coming soon_

#### Sample:Sample Scatter Plots
![Bismark Scatter Plot](../../examples/bismark_methylation_scatter.png)

#### Methylation Score Histograms
<table>
<tr><th>All Samples</th><th>Per-Sample</th></tr>
<tr><td>
<img src="../../examples/bismark_histogram_allsamples.png" title="Histogram - All Samples">
</td><td>
<img src="../../examples/bismark_histogram_perSample.png" title="Histogram - Per Samples">
</td></tr></table>

#### Coverage Analysis
_Example coming soon_

#### Principle Component Analysis
_Example possibly coming soon. Maybe._

## Usage

On the command line:
```bash
python bismark.py *cov
```

Within a python script:

```python
from ngi_visualizations import bismark
bismark.run_comparisons(input_gwCov_file_paths)
```

## Dependencies

These scripts are written in Python and run on the command line. They require
output generated by [Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/).

The following Python libraries are required:

* [matplotlib](http://matplotlib.org/)
* [scipy](http://scipy.org/)
* argparse
* logging
* os

## Credits
These scripts were written for use at the
[National Genomics Infrastructure](https://portal.scilifelab.se/genomics/)
at [SciLifeLab](http://www.scilifelab.se/) in Stockholm, Sweden. They are
part of a larger repository of
[NGI Visualization Scripts](https://github.com/SciLifeLab/ngi_visualizations).

For more information, please get in touch with
[Phil Ewels](https://github.com/ewels).

<p align="center"><a href="http://www.scilifelab.se/" target="_blank"><img src="../../examples/SciLifeLab_logo.png" title="SciLifeLab"></a></p>
