# Qualimap Plots

[Qualimap](http://qualimap.bioinfo.cipf.es/) is a tool to
_"facilitate the quality control of alignment sequencing data"_.

Whilst it produces nice plots out of the box, we have created some scripts
to re-plot some of the data to give a more consistent feel to the output.
We also split some of the plots up into per-chromosome graphs to give
a better view of the data.

## Coverage Histogram
Replacement plot for the Qualimap
[Coverage Histogram (0-50X)](http://kokonech.github.io/qualimap/HG00096.chr20_bamqc/images_qualimapReport/genome_coverage_0to50_histogram.png).
Essentially just stylistic changes - larger axis labels, different colours,
different aspect ratio and so on.

![Coverage Histogram](../examples/qualimap_coverage.png)

## Genome Fraction Coverage
Replacement plot for the Qualimap
[Genome Fraction Coverage Histogram](http://kokonech.github.io/qualimap/HG00096.chr20_bamqc/images_qualimapReport/genome_coverage_quotes.png).

![Insert Size Histogram](../examples/genome_fraction.png)

## GC Distribution
Replacement plot for the Qualimap
[Mapped Reads GC-contentent Distribution Plot](http://kokonech.github.io/qualimap/HG00096.chr20_bamqc/images_qualimapReport/genome_gc_content_per_window.png).

![Insert Size Histogram](../examples/gc_distribution.png)

## Usage
These scripts work on the command line and take the raw Qualimap output as
inputs:

    coverage_histogram.py coverage_histogram.txt
    insert_size.py insert_size_histogram.txt
    genome_fraction_coverage.py genome_fraction_coverage.txt
    gc_distribution.py mapped_reads_gc-content_distribution.txt

All of these `.txt` files are generated
by [Qualimap](http://qualimap.bioinfo.cipf.es/) and can typically be found in
`raw_data_qualimapReport` report output folder.


## Parameters
Parameters are very similar, so shown in one table. Defaults are for the
coverage histogram and insert size histograms, respectively.

Command Line Flag | Description
----------------- | -------------------- | -----------
`<input>` | Required.<br>Qualimap output. See above.
`-o`, `--output` | Default: _varies_<br>Plot output filename base. 
`-x`, `--min_x` | Default: `0` or `None`<br>Minimum x limit. `None` to use data limit.
`-m`, `--max_x` | Default: `60` or `None`<br>Maximum x limit. `None` to use data limit.
`-b`, `--bin_size` | Default: `10`<br>Bin size for histogram.  Insert Size Histogram only.
`-l`, `--log` | Default: `info`<br>Level of log messages to display. Can be `debug`, `info` or `warning`.
`-u`, `--log-output` | Default: `stdout`<br>Log output filename.
`-h`, `--help` | Display the help.

## Dependencies

These scripts are written in Python and run on the command line. They require
output generated by [Qualimap](http://qualimap.bioinfo.cipf.es/).

The following Python libraries are required:

* [matplotlib](http://matplotlib.org/)
* argparse
* logging
* os

## Credits
These scripts were written for use at the 
[National Genomics Infrastructure](https://portal.scilifelab.se/genomics/)
at [SciLifeLab](http://www.scilifelab.se/) in Stockholm, Sweden. They are 
part of a larger repository of
[Visualization Scripts](https://github.com/SciLifeLab/visualizations).

For more information, please get in touch with
[Phil Ewels](https://github.com/ewels).

<p align="center"><a href="http://www.scilifelab.se/" target="_blank"><img src="../examples/SciLifeLab_logo.png" title="SciLifeLab"></a></p>
